### PR TITLE
Make deferred messages queues configurable

### PR DESCRIPTION
**Pull request by Liorgal28**
This PR inserts 2 main changes:
1) Make the deferred messages queue size configurable.
2) Decrease the num of TLS threads to 1 in order to avoid thread race.
3) Fix Asan heap-use-after-free in post-execution handle of the deferred message.

### CODE REVIEW

**Review by yontyon — APPROVED** [review]

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 1]
```diff
@@ -1043,6 +1045,7 @@ void ReplicaImp::onMessage<PrePrepareMsg>(PrePrepareMsg *msg) {
     LOG_INFO(GL,
              "Ignoring PrePrepareMsg because system is stopped at checkpoint pending control state operation (upgrade, "
              "etc...)");
+    // TODO(LG): we should delete the message.
```
Comment (robem) [inline bftengine/src/bftengine/ReplicaImp.cpp:1048]: Then why don't we do it?

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 2]
```diff
@@ -1078,9 +1081,14 @@ void ReplicaImp::onMessage<PrePrepareMsg>(PrePrepareMsg *msg) {
 
     return;  // TODO(GG): memory deallocation is confusing .....
   }
+  if (isCurrentPrimary() && (msg->senderId() != getReplicaConfig().replicaId)) {
+    LOG_INFO(GL, "Ignoring PrePrepareMsg since im the current primary");
```
Comment (robem) [inline bftengine/src/bftengine/ReplicaImp.cpp:1085]: "... since this is the current primary" or "... since I'm the current primary"

**Review by robem — COMMENTED** [review]
Where are the commit messages? What problem are you solving and how?

File: communication/src/TlsTCPCommunication.cpp [hunk 3]
```diff
@@ -14,7 +14,7 @@
 #include "TlsRunner.h"
 
 // TODO: Make this configurable
-static constexpr size_t NUM_THREADS = 2;
+static constexpr size_t NUM_THREADS = 1;
```
Comment (robem) [inline communication/src/TlsTCPCommunication.cpp:17]: Why is this constant being changed?

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 1 above)
Comment (Liorgal28) [reply bftengine/src/bftengine/ReplicaImp.cpp]: You're right it shouldn't be in this PR - removed.

**Pull request merged**